src/node/transaction.cpp

```diff
@@ -28,65 +28,83 @@ static TransactionError HandleATMPError(const TxValidationState& state, std::str
 {
-    // BroadcastTransaction can be called by either sendrawtransaction RPC or wallet RPCs.
-    // node.peerman is assigned both before chain clients and before RPC server is accepting calls,
-    // and reset after chain clients and RPC sever are stopped. node.peerman should never be null here.
-    assert(node.peerman);
+    // BroadcastTransaction can be called by either sendrawtransaction RPC or the wallet.
+    // chainman, mempool and peerman are initialized before the RPC server and wallet are started
+    // and reset after the RPC sever and wallet are stopped.
+    assert(node.chainman);
     assert(node.mempool);
+    assert(node.peerman);
+
     std::promise<void> promise;
-    uint256 hashTx = tx->GetHash();
+    uint256 txid = tx->GetHash();
+    uint256 wtxid = tx->GetWitnessHash();
     bool callback_set = false;
 
-    { // cs_main scope
-    assert(node.chainman);
-    LOCK(cs_main);
-    // If the transaction is already confirmed in the chain, don't do anything
-    // and return early.
-    CCoinsViewCache &view = node.chainman->ActiveChainstate().CoinsTip();
-    for (size_t o = 0; o < tx->vout.size(); o++) {
-        const Coin& existingCoin = view.AccessCoin(COutPoint(hashTx, o));
-        // IsSpent doesn't mean the coin is spent, it means the output doesn't exist.
-        // So if the output does exist, then this transaction exists in the chain.
-        if (!existingCoin.IsSpent()) return TransactionError::ALREADY_IN_CHAIN;
-    }
-    if (!node.mempool->exists(hashTx)) {
-        // Transaction is not already in the mempool.
-        if (max_tx_fee > 0) {
-            // First, call ATMP with test_accept and check the fee. If ATMP
-            // fails here, return error immediately.
+    {
+        LOCK(cs_main);
+
+        // If the transaction is already confirmed in the chain, don't do anything
+        // and return early.
+        CCoinsViewCache &view = node.chainman->ActiveChainstate().CoinsTip();
+        for (size_t o = 0; o < tx->vout.size(); o++) {
+            const Coin& existingCoin = view.AccessCoin(COutPoint(txid, o));
+            // IsSpent doesn't mean the coin is spent, it means the output doesn't exist.
+            // So if the output does exist, then this transaction exists in the chain.
+            if (!existingCoin.IsSpent()) return TransactionError::ALREADY_IN_CHAIN;
+        }
+
+        if (auto mempool_tx = node.mempool->get(txid); mempool_tx) {
+            // There's already a transaction in the mempool with this txid. Don't
+            // try to submit this transaction to the mempool (since it'll be
+            // rejected as a TX_CONFLICT), but do attempt to reannounce the mempool
+            // transaction if relay=true.
+            //
+            // The mempool transaction may have the same or different witness (and
+            // wtxid) as this transaction. Use the mempool's wtxid for reannouncement.
+            wtxid = mempool_tx->GetWitnessHash();
+        } else {
```

Comment:
Maybe I am confused and missing other details, but here's my doubt.

It seems here that if a new tx (with a same mempool-tx `txid` but diff `wtxid`) is passed into broadcast, then it will just check against the `txid`, see that it already exist in mempool, and then attempt to broadcast the existing `wtxid` instead of the new `wtxid`. It also wont try to add it to mempool or the unbroadcast list. But it should, as its a brand new transaction.

It seems to me that in case of `same-txid-different-wtxid` here, nothing happens with the new transaction, and we just relay the existing tx and call it a day? How is that not a problem?

Reply:
> But it should, as its a brand new transaction.

In terms of adding it to the mempool, it cannot since it'll be rejected as `TX_CONFLICT`. And if we cannot add it to our mempool, then we should not relay the new `wtxid` as this is not in our mempool.

AFAIU with `same-txid-different-wtxid` the effect of the transaction (in terms of inputs and outputs) would be identical as it's just the witness that would be different. I don't think it would matter which one ends up being confirmed.

At least this is how I understand it to be :)

Reply:
Thanks @duncandean, that makes sense. So in the case of `same-txid-different-wtxid` transaction, we are really talking about the same transaction (input-output set) but for some reason with a different signature value.

In that case it would make sense to only handle one version of it, and yes its better to handle the version already in the mempool.